test/tlsconnection.js

```diff
@@ -0,0 +1,1160 @@
+        sinon.spy(crypto.subtle, 'encrypt');
+        try {
+          await server.send(TEST_VECTORS.SERVER_RAW_APP_DATA);
+          assert.equal(SERVER_SENT.length, 4);
+          assert.ok(bytesAreEqual(SERVER_SENT[3], TEST_VECTORS.SERVER_APP_DATA));
+          assert.equal(crypto.subtle.encrypt.callCount, 1);
+        } finally {
+          crypto.subtle.encrypt.restore();
+        }
+      });
+
+      describe('handles first exchange of application data, and then', () => {
+        beforeEach(async () => {
+          const data = await server.recv(TEST_VECTORS.CLIENT_APP_DATA);
+          assert.ok(bytesAreEqual(data, TEST_VECTORS.CLIENT_RAW_APP_DATA));
+          await server.send(TEST_VECTORS.SERVER_RAW_APP_DATA);
+          assert.equal(SERVER_SENT.length, 4);
+          assert.ok(bytesAreEqual(SERVER_SENT[3], TEST_VECTORS.SERVER_APP_DATA));
+        });
+
+        describe('handles second exchange of application data, and then', () => {
+          beforeEach(async () => {
+            const data = await server.recv(TEST_VECTORS.CLIENT_APP_DATA_2);
+            assert.ok(bytesAreEqual(data, TEST_VECTORS.CLIENT_RAW_APP_DATA_2));
+            await server.send(TEST_VECTORS.SERVER_RAW_APP_DATA_2);
+            assert.equal(SERVER_SENT.length, 5);
+            assert.ok(bytesAreEqual(SERVER_SENT[4], TEST_VECTORS.SERVER_APP_DATA_2));
+          });
+
+          describe('accepts an explicit close alert from the client, and then', () => {
+            beforeEach(async () => {
+              const data = await server.recv(TEST_VECTORS.CLIENT_CLOSE);
+              assert.equal(data, null);
+            });
+
+            it('does not accept any additional received data', async () => {
+              // Forge a valid record, so we know it fails because of the close, rather than crypto error.
+              const cipherstate = await testHelpers.makeCipherState(server._keyschedule.clientApplicationTrafficSecret, 3);
+              await assert.throwsAsync(async () => {
+                await server.recv(await testHelpers.makeEncryptedRecord(cipherstate, {}));
+              }, TLSError, 'CLOSE_NOTIFY');
+            });
+
+            describe('is able to send an explicit close in return, and then', () => {
+              beforeEach(async () => {
+                await server.close();
+                assert.equal(SERVER_SENT.length, 6);
+                assert.ok(bytesAreEqual(SERVER_SENT[5], TEST_VECTORS.SERVER_CLOSE));
+              });
+
+              it('rejects any further attempts to send data', async () => {
+                await assert.throwsAsync(async () => {
+                  await server.send(TEST_VECTORS.SERVER_RAW_APP_DATA);
+                }, TLSError, 'CLOSE_NOTIFY');
+              });
+            });
+          });
+        });
+
+        it('rejects an out-of-order close notification', async () => {
+          await assert.throwsAsync(async () => {
+            await server.recv(TEST_VECTORS.CLIENT_CLOSE);
```

Comment:
I have to admit to having trouble following the events that lead to this point.  I think that you need a comment explaining where you are and why this fails.